pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -0,0 +1,344 @@
+type DockercfgControllerOptions struct {
+	// Resync is the time.Duration at which to fully re-list service accounts.
+	// If zero, re-list will be delayed as long as possible
+	Resync time.Duration
+}
+
+// NewDockercfgController returns a new *DockercfgController.
+func NewDockercfgController(cl client.Interface, options DockercfgControllerOptions) *DockercfgController {
+	e := &DockercfgController{
+		client: cl,
+	}
+
+	_, e.serviceAccountController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return e.client.ServiceAccounts(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return e.client.ServiceAccounts(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.ServiceAccount{},
+		options.Resync,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    e.serviceAccountAdded,
+			UpdateFunc: e.serviceAccountUpdated,
+		},
+	)
+
+	e.dockerURL = OpenshiftDockerURL
+
+	return e
+}
+
+// DockercfgController manages dockercfg secrets for ServiceAccount objects
+type DockercfgController struct {
+	stopChan chan struct{}
+
+	client client.Interface
+
+	dockerURL     string
+	dockerURLLock sync.Mutex
+
+	serviceAccountController *framework.Controller
+}
+
+// Runs controller loops and returns immediately
+func (e *DockercfgController) Run() {
+	if e.stopChan == nil {
+		e.stopChan = make(chan struct{})
+		go e.serviceAccountController.Run(e.stopChan)
+	}
+}
+
+// Stop gracefully shuts down this controller
+func (e *DockercfgController) Stop() {
+	if e.stopChan != nil {
+		close(e.stopChan)
+		e.stopChan = nil
+	}
+}
+
+func (e *DockercfgController) SetDockerURL(newDockerURL string) {
+	e.dockerURLLock.Lock()
+	defer e.dockerURLLock.Unlock()
+
+	e.dockerURL = newDockerURL
+}
+
+// serviceAccountAdded reacts to a ServiceAccount creation by creating a corresponding ServiceAccountToken Secret
+func (e *DockercfgController) serviceAccountAdded(obj interface{}) {
+	serviceAccount := obj.(*api.ServiceAccount)
+
+	if err := e.createDockercfgSecretIfNeeded(serviceAccount); err != nil {
+		glog.Error(err)
+	}
+}
+
+// serviceAccountUpdated reacts to a ServiceAccount update (or re-list) by ensuring a corresponding ServiceAccountToken Secret exists
+func (e *DockercfgController) serviceAccountUpdated(oldObj interface{}, newObj interface{}) {
+	newServiceAccount := newObj.(*api.ServiceAccount)
+
+	if err := e.createDockercfgSecretIfNeeded(newServiceAccount); err != nil {
+		glog.Error(err)
+	}
+}
+
+const (
+	dockercfgToken = "-dockercfg-"
+
+	// These constants are here to create a name that is short enough to survive chopping by generate name
+	maxNameLength               = 63
+	randomLength                = 5
+	maxServiceAccountNameLength = maxNameLength - randomLength - len(dockercfgToken)
+)
+
+func getUsableServiceAccountNamePrefix(serviceAccountName string) string {
```

Comment:
cesar did something similar to get deterministic non-conflicting pod names for builds and deployments... if the name was going to overflow, he chopped it short and included a 8-char hash of the name. It's a vanishingly small scenario, but I thought I'd mention the approach.